src/java.base/share/classes/java/util/concurrent/ForkJoinPool.java

```diff
@@ -1835,10 +1835,11 @@ private int scan(WorkQueue w, int prevSrc, int r) {
                 else if (t != null && WorkQueue.casSlotToNull(a, k, t)) {
                     q.base = nb;
                     w.source = src;
-                    if (prevSrc == 0 && q.base == nb && a[nk] != null)
-                        signalWork();           // propagate
+                    if (src + (src << SWIDTH) != prevSrc &&
+                        q.base == nb && a[nk] != null)
+                        signalWork();           // propagate at most twice/run
                     w.topLevelExec(t, q);
-                    return src;
+                    return src + (prevSrc << SWIDTH);
```

Comment:
Are all of these values (`src`, `prevSrc`) less or equal `0x7FFF`? Otherwise we are playing with sign bits here.

Reply:
Yes, see the discussions around line 391 of internal docs.